dimod/views/quadratic.py

```diff
@@ -390,3 +395,90 @@ def set_linear(self, v: Variable, bias: Bias):
     @abc.abstractmethod
     def set_quadratic(self, u: Variable, v: Variable, bias: Bias):
         raise NotImplementedError
+
+    def iter_linear(self) -> Iterator[Tuple[Variable, Bias]]:
+        """Iterate over the variables and their biases."""
+        get = self.get_linear
+        for v in self.variables:
+            yield v, get(v)
+
+    def to_polystring(self, encoder: Optional[Callable[[Variable], str]] = None) -> str:
+        """Return a string representing the model as a polynomial.
+
+        Args:
+            encoder: A function mapping variables to a string. By default
+                string variables are mapped directly whereas all other types
+                are mapped to a string :code:`f"v{variable!r}"`.
+
+        Returns:
+            A string representing the binary quadratic model.
+
+        Examples:
+
+            >>> x, y, z = dimod.Binaries(['x', 'y', 'z'])
+            >>> (2*x + 3*y*z + 6).to_polystring()
+            '6 + 2*x + 3*y*z'
+
+        """
+
+        if encoder is None:
+            def encoder(v: Variable) -> str:
+                return v if isinstance(v, str) else f"v{v!r}"
+
+        # developer note: we use floats everywhere because they have some nice
+        # methods and because for this method we're not too worried about
+        # performance
+
+        def neg(bias: float) -> bool: return bias < 0
+
+        def string(bias: float) -> str:
+            return repr(abs(int(bias))) if bias.is_integer() else repr(abs(bias))
+
+        def coefficient(bias: float) -> str:
+            return '' if abs(bias) == 1 else f"{string(bias)}*"
+
+        # linear variables that have a positive bias or are not going to be
+        # included implicitly in quadratic
+        linear = ((v, float(bias)) for v, bias in self.iter_linear()
+                  if bias or not any(bias for _, bias in self.iter_neighborhood(v)))
```

Comment:
I guess you want to show all variables in the model but (1) typical polynomial representations do not include variables with zero coefficients and (2) if the string is used as input to some parsing code, it complicates it by having linear terms appear/disappear dependent on a quadratic bias (e.g., the linear z term in `'-x + 0*y + 0*z'` versus `'-x + 0*y - x*z'`)

Reply:
(1) I am not sure there is a "typical"
(2) We already handle this case in BQM/QM construction. E.g. `BQM({'x': -1, 'y': 0}, {'xz': -1}, 0.0, 'BINARY')` implicitly adds 'z'. That's why we only ignore linear variables when they appear in a quadratic term.

Reply:
I mean a mathematical representation rather than a printout of the QM data structure. My impression is that the intention here is to print the math format of a polynomial rather than a string with tided-up values of qm.linear  + qm.quadratic. Is that a mistaken understanding?

Reply:
I am saying that there is no "math format". Writing a polynomial `E(x, y) = 2*x` or `E(x, y) = 2*x + 0*y` are both perfectly valid. I agree the latter looks a bit more redundant, but IMO is is better to err on the side of [explicitly](https://www.python.org/dev/peps/pep-0020/#id2) including a variable rather than droping it implicitly.

Reply:
To be fair - this case only arises with isolated variables with 0 linear bias. So really shouldn't show up in most use cases unless something is going wrong.

Reply:
The usecase that comes to my mind is someone preloading variable labels, then in iterations adding biases/interactions and printing out the polynomial for verification, as part of self-checking or frontal training.
But that might not be the kind of usecase this was added for. No worries.

Reply:
I think that case it's even more important. For example:
```python
>>> bqm = BQM.empty('SPIN')
>>> bqm.offset = 1
>>> bqm.to_polystring()
'1'
>>> bqm.add_variable('a')
>>> bqm.to_polystring()
'1'
```
seems confusing to me. Whereas in
```python
>>> bqm.add_variable('a')
>>> bqm.to_polystring()
'1 + 0*a'
```
I can see that a change has happened.

Reply:
That's a good usecase.
The one I had in mind is more like this:

```
>>> qm = QuadraticModel()
>>> qm.add_variables_from('INTEGER', ['a', 'b', 'c'])
>>> qm.to_polystring()
''
>>> qm.add_linear('a', -1)
>>> qm.to_polystring()
'-a'
>>> qm.add_linear('b', 2)
'-a + 2*b'
```